aip/auth/4120.md

```diff
@@ -0,0 +1,96 @@
+ - Credentials accept a user-supplied `universe_domain`.
+ - Service Account and External Account credentials read the universe domain from the credentials JSON file.
+ - GCE credentials retrieve the universe domain from the metadata server.
+
+### Naming
+
+We aim to have a consistent user experience across a broad range of clients. One of the invariants is the `universe_domain` name that should be used to represent the universe domain in configuration, user-readable properties and methods. Languages that use different casing conventions should adapt the name to be idiomatic but keep the two specific words in the same order, so `universe-domain`, `universeDomain` and `SetUniverseDomain` are all acceptable, but e.g. `DomainOfUniverse` or `Endpoint` is not.
+
+### Credentials implement a `universe_domain` getter (method, property)
+
+> This requirement applies to **all** flows.
+
+Every credentials class/object **must** expose the `universe_domain` getter (property, method, etc). The auth library end-users, including the client libraries will use this property.
+
+The credentials classes for flows that are not supported outside the GDU (e.g. user credentials) **must** still expose the `universe_domain` to the end-user. These credentials **must** return the default value `"googleapis.com"` for the `universe_domain`.
+
+### Credentials accept a user-supplied `universe_domain`
+
+> This requirement applies **only** to the flows supported outside the GDU.
+
+The end-user **must** be able to create universe-specific credentials in code. This can take many forms: constructor arguments, set properties/methods, copy-on-write methods, etc. This requirement is to treat `universe_domain` as a first-class property in regards to modification and user visibility.
+
+E.g. if the authentication library provides a `with_scope` copy-on-write method it can satisfy this requirement by providing a similar `with_universe_domain` method.
+This requirement applies only to the credentials classes that are related to the authentication flows supported outside the GDU. This is in contrast to the requirement to "expose the `universe_domain` to the end-user" which applies to all credentials classes.
+
+### Service Account credentials read the universe domain from the credentials JSON file
+
+> This requirement applies **only** to the Service Account flow.
+
+The JSON credential files will have a new property: `universe_domain`. The authentication libraries **must** read it when parsing JSON credential files and expose it to the libraries' end-users. If the `universe_domain` is not present, the credentials file is assumed to be (a legacy one) from the Google default universe, and the `universe_domain` value **must** default to `"googleapis.com"`.
+
+This value can also be used to make decisions about whether the library is in the GDU or not e.g. the purpose of throwing specific errors if the end-user is forcing an unsupported authentication flow with non-GDU credentials.
```

Comment:
I think this must be true for all credential types, so maybe move it to the "Credentials implement a `universe_domain` getter (method, property)" section?

Reply:
Good idea! done.